lib/source/pl/core/preprocessor.cpp

```diff
@@ -131,77 +134,109 @@ namespace pl::core {
+                m_defines[name].clear();
+                m_defines[name] = values;
+                removeValue(m_keys, name);
+                m_keys.emplace_back(tokenValue);
+            }
+        } else {
+            m_defines[name] = values;
+            m_keys.emplace_back(tokenValue);
+        }
+    }
 
-        auto value = getDirectiveValue(true);
+    void Preprocessor::handleUnDefine(unsigned line) {
+        auto token = getDirectiveValue(line);
 
-        if(!value.has_value()) {
-            error("Expected value after #define");
+        if(!isNameValid(token)) {
+            error("Expected identifier after #undef");
             return;
         }
-
-        m_defines[name.value()] = { value.value(), m_lineNumber };
+        auto name = getOptionalValue(token);
+        if (m_defines.contains(name)) {
+            m_defines.erase(name);
+            removeValue(m_keys, name);
+        }
     }
 
-    void Preprocessor::handlePragma() {
-        auto key = getDirectiveValue();
+    void Preprocessor::handlePragma(unsigned line) {
+        auto token = getDirectiveValue(line);
 
-        if(!key.has_value()) {
+        if(!token.has_value()) {
             errorDesc("No instruction given in #pragma directive.", "A #pragma directive expects a instruction followed by an optional value in the form of #pragma <instruction> <value>.");
             return;
         }
-
-        auto value = getDirectiveValue(true);
-
-        this->m_pragmas[key.value()].emplace_back(value.value_or(""), m_lineNumber);
+        auto key = getOptionalValue(token);
+        token = getDirectiveValue(line);
+        if (!token.has_value())
+            this->m_pragmas[key].emplace_back("", line);
+        else
+            this->m_pragmas[key].emplace_back(getOptionalValue(token), line);
     }
 
-    void Preprocessor::handleInclude() {
-        const auto includeFile = getDirectiveValue();
+    void Preprocessor::handleInclude(unsigned line) {
+        auto token = getDirectiveValue(line);
+        if (!token.has_value()) {
+            errorDesc("No file to include given in #include directive.", "A #include directive expects a path to a file: #include \"path/to/file\" or #include <path/to/file>.");
+            return;
+        }
+        auto includeFile = getOptionalValue(token);
 
-        if (!includeFile.has_value()) {
+        if (includeFile.empty()){
             errorDesc("No file to include given in #include directive.", "A #include directive expects a path to a file: #include \"path/to/file\" or #include <path/to/file>.");
             return;
         }
 
-        if (includeFile->starts_with('"') && includeFile->ends_with('"'))
-            ; // Parsed path wrapped in ""
-        else if (includeFile->starts_with('<') && includeFile->ends_with('>'))
-            ; // Parsed path wrapped in <>
-        else {
+        if ((!includeFile.starts_with('"') || !includeFile.ends_with('"')) &&
```

Comment:
I feel like this could be simplified using de-morgan

Reply:
I made this statement the simplest I could.